File: sol/solution/0001-0100/0002. Add Two Numbers/Solution.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).
-spec add_two_numbers(L1 :: #list_node{} | null, L2 :: #list_node{} | null) -> #list_node{} | null.
add_two_numbers(L1, L2) ->
    add_two_numbers(L1, L2, 0, #list_node{val = 0}).
add_two_numbers(null, null, 0, Dummy) ->
    Dummy#list_node.next;
add_two_numbers(L1, L2, Carry, Dummy) ->
    Sum = Carry + 
          case L1 of
              null -> 0;
              _ -> L1#list_node.val
          end +
          case L2 of
              null -> 0;
              _ -> L2#list_node.val
          end,
    NewCarry = Sum div 10,
    NewNode = #list_node{val = Sum rem 10},
    NextL1 = case L1 of
                 null -> null;
                 _ -> L1#list_node.next
             end,
    NextL2 = case L2 of
                 null -> null;
                 _ -> L2#list_node.next
             end,
    case L1 =:= null andalso L2 =:= null andalso NewCarry =:= 0 of
        true -> NewNode;
        false -> NewNode#list_node{next = add_two_numbers(NextL1, NextL2, NewCarry, #list_node{val = 0})}
    end.

File: sol/solution/0001-0100/0002. Add Two Numbers/Solution.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end
defmodule Solution do
  @spec add_two_numbers(l1 :: ListNode.t | nil, l2 :: ListNode.t | nil) :: ListNode.t | nil
  def add_two_numbers(l1, l2) do
    add_two_numbers(l1, l2, 0)
  end
  defp add_two_numbers(nil, nil, 0), do: nil
  defp add_two_numbers(l1, l2, carry) do
    val1 = if l1, do: l1.val, else: 0
    val2 = if l2, do: l2.val, else: 0
    sum = val1 + val2 + carry
    new_carry = div(sum, 10)
    next1 = if l1, do: l1.next, else: nil
    next2 = if l2, do: l2.next, else: nil
    %ListNode{
      val: rem(sum, 10),
      next: add_two_numbers(next1, next2, new_carry)
    }
  end
end